concent_api/core/payments/sci_callback.py

```diff
@@ -0,0 +1,260 @@
+        )
+
+    # If the received Golem message is anything else than SignedTransaction, TransactionRejected or and ERROR frame.
+    if not isinstance(deserialized_message.payload, SignedTransaction):
+        raise SCICallbackPayloadError('Received frame payload is not Golem message SignedTransaction instance.')
+
+    # If the received Golem message is not signed by the Signing Service.
+    try:
+        deserialized_message.payload.verify_signature(settings.SIGNING_SERVICE_PUBLIC_KEY)
+    except InvalidSignature as exception:
+        raise SCICallbackPayloadSignatureError() from exception
+
+    return deserialized_message.payload
+
+
+def verify_data_and_signature(signed_transaction: SignedTransaction, transaction: Transaction) -> None:
+    """
+    Verify received SignedTransaction data and signature and handle related errors.
+
+    1. Check if received SignedTransaction data match original Ethereum Transacion.
+    2. Check if received signature is created for given message with given keys pair.
+    """
+
+    assert isinstance(signed_transaction, SignedTransaction)
+    assert isinstance(transaction, Transaction)
+
+    # The transaction data in SignedTransaction does not match the information we sent.
+    if not is_signed_transaction_data_equal_to_transaction_data(signed_transaction, transaction):
+        raise SCICallbackPayloadError('Received SignedTransaction does not match TransactionSigningRequest.')
+
+    # If the Ethereum signature in SignedTransaction is not a valid signature or
+    # or the transaction in SignedTransaction is not signed with the right Ethereum private key.
+    message_hash = sha3(encode(transaction, UnsignedTransaction))
+
+    if not ecdsa_raw_verify(
+        message_hash,
+        (
+            signed_transaction.v,
+            signed_transaction.r,
+            signed_transaction.s,
+        ),
+        settings.CONCENT_ETHEREUM_PUBLIC_KEY,
+    ):
+        raise SCICallbackTransactionSignatureError(
+            'Received SignedTransaction signature data is not signed by right Ethereum private key.'
+        )
+
+
+def is_signed_transaction_data_equal_to_transaction_data(
+    signed_transaction: SignedTransaction,
+    transaction: Transaction
+) -> bool:
+    """ Compare SignedTransaction message data with Ethereum Transaction data. """
+
+    assert isinstance(signed_transaction, SignedTransaction)
+    assert isinstance(transaction, Transaction)
+
+    return (
+        signed_transaction.nonce == transaction.nonce and
+        signed_transaction.gasprice == transaction.gasprice and
+        signed_transaction.startgas == transaction.startgas and
+        signed_transaction.to == transaction.to and
+        signed_transaction.value == transaction.value and
+        signed_transaction.data == transaction.data
+    )
+
+
+def copy_transaction_and_signature(signed_transaction: SignedTransaction, transaction: Transaction) -> Transaction:
```

Comment:
as stated above (where the function is called): please rename it

Reply:
:+1: